lib/prelude/input.dart

```diff
@@ -1,14 +1,216 @@
+
+  // Representations
+  Map<int, TapInteraction> tapInteractions = {};
+  Map<int, DragInteraction> dragInteractions = {};
+
+  /// Get the taps and long taps which started this frame
+  List<TapInteraction> newTaps() {
+    final result = <TapInteraction>[];
+    for (final tap in taps) {
+      if (tapInteractions.containsKey(tap.pointerId)) {
+        result.add(tapInteractions[tap.pointerId]!);
+      }
+    }
+    for (final longTap in longTaps) {
+      if (tapInteractions.containsKey(longTap.pointerId)) {
+        result.add(tapInteractions[longTap.pointerId]!);
+      }
+    }
+    return result;
+  }
+
+  /// Get the taps and long taps which ended this frame
+  List<TapInteraction> newTapUps() {
+    final result = <TapInteraction>[];
+    for (final tapUp in tapUps) {
+      if (tapInteractions.containsKey(tapUp.pointerId)) {
+        result.add(tapInteractions[tapUp.pointerId]!);
+      }
+    }
+    return result;
+  }
+
+  /// Get the taps and long taps which were cancelled this frame
+  List<TapInteraction> newTapCancels() {
+    final result = <TapInteraction>[];
+    for (final tapCancel in tapCancels) {
+      if (tapInteractions.containsKey(tapCancel.pointerId)) {
+        result.add(tapInteractions[tapCancel.pointerId]!);
+      }
+    }
+    return result;
+  }
+
+  /// Get the drags which started this frame
+  List<DragInteraction> newDragStarts() {
+    final result = <DragInteraction>[];
+    for (final dragStart in dragStarts) {
+      if (dragInteractions.containsKey(dragStart.pointerId)) {
+        result.add(dragInteractions[dragStart.pointerId]!);
+      }
+    }
+    return result;
+  }
+
+  /// Get the drags which updated this frame
+  List<DragInteraction> newDragUpdates() {
+    final result = <DragInteraction>[];
+    for (final dragUpdate in dragUpdates) {
+      if (dragInteractions.containsKey(dragUpdate.pointerId)) {
+        result.add(dragInteractions[dragUpdate.pointerId]!);
+      }
+    }
+    return result;
+  }
+
+  /// Get the drags which ended this frame
+  List<DragInteraction> newDragEnds() {
+    final result = <DragInteraction>[];
+    for (final dragEnd in dragEnds) {
+      if (dragInteractions.containsKey(dragEnd.pointerId)) {
+        result.add(dragInteractions[dragEnd.pointerId]!);
+      }
+    }
+    return result;
+  }
+
+  void clear() {
+    keysPressed = {};
+    taps = [];
+    longTaps = [];
+    tapUps = [];
+    tapCancels = [];
+    dragStarts = [];
+    dragUpdates = [];
+    dragEnds = [];
+  }
+
+  void process(double dt) {
+    // Set last frame position for drags
+    for (final interaction in dragInteractions.values) {
+      interaction.positionAtLastFrame = interaction.position;
+    }
+
+    // Use the events from this frame to update the pointer interactions
+    // -- normal taps
+    for (final tap in taps) {
+      assert(tapInteractions.containsKey(tap.pointerId) == false,
+          'Pointer ID ${tap.pointerId} already exists');
```

Comment:
I think the assert message is wrong, the point is not a key of the map as you check for false?

Reply:
If I understand asserts correctly, if the expression is not `true` it throws an exception with the message. In this case I need to validate that the key **doesn't** exist yet, otherwise it is undefined behavior which I didn't expect from Flutter/Flame.

Reply:
You check if a key is not present, and the text says that the key already exists. I think it's either wrong or super confusing, or I'm too special to get it :)